src/System.Text.Json/src/System/Text/Json/JsonParseObject.cs

```diff
@@ -0,0 +1,350 @@
+
+            if (typeCode == (byte)JsonParser.JsonValueType.Object || typeCode == (byte)JsonParser.JsonValueType.Array)
+            {
+                int length = BitConverter.ToInt32(_buffer, _start + 4);
+                if (length == 0) return false;
+                return true;
+            }
+            else
+            {
+                int location = BitConverter.ToInt32(_buffer, _start);
+                if (_buffer[location - 1] == '"' && _buffer[location + 4] == '"')
+                {
+                    return true;
+                }
+                return (_buffer[location] != 'n' || _buffer[location + 1] != 'u' || _buffer[location + 2] != 'l' || _buffer[location + 3] != 'l');
+            }
+        }
+
+        public JsonParseObject this[string index]
+        {
+            get
+            {
+                int length = BitConverter.ToInt32(_buffer, _start + 4);
+                var typeCode = _buffer[_start + 8];
+
+                if (length == 0)
+                {
+                    throw new KeyNotFoundException();
+                }
+
+                if (typeCode != (byte)JsonParser.JsonValueType.Object)
+                {
+                    throw new NullReferenceException();
+                }
+
+                for (int i = _start + RowSize; i <= _end; i += RowSize)
+                {
+                    length = BitConverter.ToInt32(_buffer, i + 4);
+                    typeCode = _buffer[i + 8];
+
+                    if (typeCode == (byte)JsonParser.JsonValueType.Object || typeCode == (byte)JsonParser.JsonValueType.Array)
+                    {
+                        i += length * RowSize;
+                        continue;
+                    }
+
+                    int location = BitConverter.ToInt32(_buffer, i);
+                    if (isEqual(index, _buffer, location, length))
+                    {
+                        int newStart = i + RowSize;
+                        int newEnd = newStart + RowSize;
+
+                        typeCode = _buffer[newStart + 8];
+
+                        if (typeCode == (byte)JsonParser.JsonValueType.Object || typeCode == (byte)JsonParser.JsonValueType.Array)
+                        {
+                            length = BitConverter.ToInt32(_buffer, newStart + 4);
+                            newEnd = newEnd + RowSize * length;
+                        }
+
+                        return new JsonParseObject(_buffer, newStart, newEnd);
+                    }
+
+                    typeCode = _buffer[i + RowSize + 8];
+
+                    if (typeCode != (byte)JsonParser.JsonValueType.Object && typeCode != (byte)JsonParser.JsonValueType.Array)
+                    {
+                        i += RowSize;
+                    }
+                }
+
+                throw new KeyNotFoundException();
+
+            }
+        }
+
+        private bool isEqual(string str, byte[] buffer, int location, int length)
```

Comment:
isEqual does not follow our naming/casing conventions. I would call it simply Equals. 
I think it would also be useful to make it clearer that the location and length refer to the array. Possibly by describing it in a comment. Lastly, we call such parameters "index" and "count", not "location" and "length"